titan-compiler/coder.lua

```diff
@@ -476,34 +476,60 @@ end
 -- @records
 --
 
--- Compute the layout for the gc fields stored in the UValue array.
--- We use a C struct to compute the layout of fields stored in mem part.
+-- Compute the record layout. We store the gc fields in UValue array and
+-- primitive fields in the mem part.
 local function rec_compute_layout(rec)
-    local index = 0
+    local gc_index = 0
+    local prim_index = 0
     local pos = {}
     for _, field in ipairs(rec.field_decls) do
         local typ = rec._field_types[field.name]
         if types.is_gc(typ) then
-            pos[field.name] = index
-            index = index + 1
+            pos[field.name] = gc_index
+            gc_index = gc_index + 1
+        else
+            pos[field.name] = prim_index
+            prim_index = prim_index + 1
         end
     end
-    return { pos = pos, size = index }
+    return { pos = pos, gc_size = gc_index, prim_size = prim_index }
```

Comment:
Could we be more defensive and have separate `gc_pos` and `prim_pos` tables instead of a shared `pos` table? This way, if there is a compiler bug and we try to use a prim field as a gc field or vice versa we get a compiler crash instead of it silently doing the wrong thing.

Reply:
Will do.